src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -466,21 +469,48 @@ adjoin_root.mk_ne_zero_of_nat_degree_lt W.monic_polynomial (X_sub_C_ne_zero y) $
   by { rw [nat_degree_polynomial, nat_degree_X_sub_C], norm_num1 }
 
 /-- The ideal $\langle X - x \rangle$ of $R[W]$ for some $x \in R$. -/
-@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := ideal.span {X_class W x}
+@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := span {X_class W x}
 
 /-- The ideal $\langle Y - y(X) \rangle$ of $R[W]$ for some $y(X) \in R[X]$. -/
-@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := ideal.span {Y_class W y}
+@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := span {Y_class W y}
+
+/-- The ideal $\langle X - x, Y - y(X) \rangle$ of $R[W]$ for some $x \in R$ and $y(X) \in R[X]$. -/
+@[simp] noncomputable def XY_ideal (x : R) (y : R[X]) : ideal W.coordinate_ring :=
+span {X_class W x, Y_class W y}
+
+-- TODO: generalise to arbitrary `y ∈ R[X]` and use `neg_polynomial`
+lemma span_polynomial_le_span_XY {x y : R} (h : W.equation x y) :
+  (span {W.polynomial} : ideal $ R[X][Y]) ≤ span {C (X - C x), Y - C (C y)} :=
+begin
+  intros _ hz,
+  rcases mem_span_singleton'.mp hz with ⟨z, rfl⟩,
+  exact mem_span_pair.mpr
+    ⟨z * (C (C (W.a₁ * y) - (X ^ 2 + C (x + W.a₂) * X + C (x ^ 2 + W.a₂ * x + W.a₄)))),
+    z * (C (C y) - (-Y - C (C W.a₁ * X + C W.a₃))),
+    by linear_combination -z * (congr_arg C $ congr_arg C $ (W.equation_iff x y).mp h)
+      with { normalization_tactic := `[rw [weierstrass_curve.polynomial], C_simp, ring1] }⟩
+end
```

Comment:
Generalized using a more conceptual proof extracted from https://github.com/leanprover-community/mathlib/blob/e9317429cd025da26e034be12a085d2be8ef4532/src/algebraic_geometry/elliptic_curve/weierstrass.lean#L471-L472

The first lemma should go under [polynomial.dvd_iff_is_root](https://leanprover-community.github.io/mathlib_docs/data/polynomial/div.html#polynomial.dvd_iff_is_root). What do you mean by "use `neg_polynomial`".
```suggestion
lemma _root_.polynomial.X_sub_C_dvd_sub_C_eval {R} (a : R) [comm_ring R] (p : R[X]) :
  (X - C a) ∣ (p - C (p.eval a)) :=
by rw [dvd_iff_is_root, is_root, eval_sub, eval_C, sub_self]

lemma span_polynomial_le_span_XY {x : R} {y : R[X]} (h : (W.polynomial.eval y).eval x = 0) :
  (span {W.polynomial} : ideal $ R[X][Y]) ≤ span {C (X - C x), Y - C y} :=
(ideal.span_singleton_le_iff_mem _).2 begin
  obtain ⟨p, hp⟩ := W.polynomial.X_sub_C_dvd_sub_C_eval y,
  obtain ⟨q, hq⟩ := dvd_iff_is_root.2 h,
  refine mem_span_pair.mpr ⟨C q, p, _⟩,
  rw [mul_comm, ← C_mul, ← hq, mul_comm, ← eq_sub_iff_add_eq', hp],
end
```

Reply:
This is also good. In my proof there's an instance of `-Y - C (C W.a₁ * X + C W.a₃)` which is just `neg_polynomial`, but this is only defined in `point.lean`. This is also used somewhere else in the file so I think the definition should be defined in this file.

Reply:
Actually we can replace `W.polynomial` by any polynomial and upgrade to iff, though the latter is a bit involved to do and I'm figuring out the best proof. Then it's essentially a statement about the kernel of the double evaluation homomorphism.

Reply:
In any case the lemma will probably remain in this file, because there aren't double polynomials elsewhere in mathlib.

Reply:
Easier than I thought :) I think both lemmas can go under [polynomial.dvd_iff_is_root](https://leanprover-community.github.io/mathlib_docs/data/polynomial/div.html#polynomial.dvd_iff_is_root), even if this one talks about double polynomial. You can restore `quotient_XY_ideal_equiv` to its original generality (`y : R[X]`) with this generalized lemma and `ideal. span_singleton_le_iff_mem`.
```
lemma mem_C_X_sub_C_Y_sub_C_iff_eval_eval_eq_zero {R} [comm_ring R]
  {x : R} {y : R[X]} {p : R[X][X]} :
  (p.eval y).eval x = 0 ↔ p ∈ span ({C (X - C x), Y - C y} : set R[X][X]) :=
begin
  rw mem_span_pair,
  split; intro h,
  { obtain ⟨q, hq⟩ := p.X_sub_C_dvd_sub_C_eval y,
    obtain ⟨r, hr⟩ := dvd_iff_is_root.mpr h,
    exact ⟨C r, q, by rw [mul_comm, ← C_mul, ← hr, mul_comm, ← eq_sub_iff_add_eq', hq]⟩ },
  { obtain ⟨q, r, rfl⟩ := h,
    simp only [eval_mul, eval_add, eval_sub, eval_C, eval_X, sub_self, mul_zero, add_zero] },
    -- or eval_simp, simp_rw [sub_self, mul_zero, add_zero]
end
```

Reply:
So should we add this to #19120?

Reply:
I added it, but I'm still unconvinced that it should belong elsewhere - I'll wait for a review